Neos.ContentRepository.Core/Classes/ContentRepository.php

```diff
@@ -94,19 +94,35 @@ public function __construct(
      */
     public function handle(CommandInterface $command): void
     {
-        // the commands only calculate which events they want to have published, but do not do the
-        // publishing themselves
-        $eventsToPublishOrGenerator = $this->commandBus->handle($command);
-
-        if ($eventsToPublishOrGenerator instanceof EventsToPublish) {
-            $eventsToPublish = $this->enrichEventsToPublishWithMetadata($eventsToPublishOrGenerator);
-            $this->eventPersister->publishEvents($this, $eventsToPublish);
-        } else {
-            foreach ($eventsToPublishOrGenerator as $eventsToPublish) {
-                assert($eventsToPublish instanceof EventsToPublish); // just for the ide
-                $eventsToPublish = $this->enrichEventsToPublishWithMetadata($eventsToPublish);
-                $this->eventPersister->publishEvents($this, $eventsToPublish);
+        $toPublish = $this->commandBus->handle($command);
+
+        if ($toPublish instanceof EventsToPublish) {
+            // simple case
+            $eventsToPublish = $this->enrichEventsToPublishWithMetadata($toPublish);
+            if ($eventsToPublish->events->isEmpty()) {
+                return;
+            }
+            $this->eventPersister->publishWithoutCatchup($eventsToPublish);
+            $this->catchupProjections();
+            return;
+        }
+
+        // control-flow aware command handling via generator
+        try {
+            $yieldedEventsToPublish = $toPublish->current();
```

Comment:
The question is why would we then even need to yield, because we might as well return one EventsToPublish then, if there is no catch up in between?! Also means e.g. pre-emptive closing of CS is pointless as other processes constraint checks will never see it when it would be useful to have.